File: src/falcon_core/physics/config/config_manipulations.py
"""Generic config manipulation functions. These work on the raw config files."""

from typing import TYPE_CHECKING

from falcon_core.physics.config.core.voltage_constraints import VoltageConstraints

from ...dependancies import np
from .core import (
    Config,
    Group,
)
from .core.adjacency import Adjacency
from .core.standard_config_connections import StandardConfigConnections
from .dependancies import (
    BarrierGate,
    BaseConnections,
    Channel,
    Gates,
    Gname,
    Impedance,
    Impedances,
    Ohmic,
    Ohmics,
    PlungerGate,
    ReservoirGate,
    ScreeningGate,
    overload,
)

if TYPE_CHECKING:
    from .dependancies import (
        BarrierGates,
        PlungerGates,
        ReservoirGates,
        ScreeningGates,
    )
    from .typing import (
        Any,
        Connection,
        UsefulGates,
        UsefulGateType,
    )

CONFIG_TO_GATE_PRIMITIVES: dict[str, "UsefulGateType"] = {
    "ScreeningGates": ScreeningGate,
    "ReservoirGates": ReservoirGate,
    "PlungerGates": PlungerGate,
    "BarrierGates": BarrierGate,
}


class ConfigManipulations:
    """Class for config manipulation functions."""

    def unpack_device_config(
        self,
        config: dict[str, "Any"],
    ) -> Config:
        """Unpacks a config object into a Config object.

        Args:
            config (dict[str, any]): The config object to unpack.

        Returns:
            Config: The unpacked config object.
        """
        groups = {}
        for key, value in config["groups"].items():
            connections = self._extract_standard_config_connections(config=value)
            order = self._extract_order(dictionary=value, connections=connections)
            new_group = Group(
                name=Channel(name=key),
                num_dots=int(value["NumDots"]),
                order=order,
                screening_gates=connections.screening_gates,
                reservoir_gates=connections.reservoir_gates,
                plunger_gates=connections.plunger_gates,
                barrier_gates=connections.barrier_gates,
            )
            groups[Gname(value=key)] = new_group
        connections = self._extract_standard_config_connections(config=config)
        ohmics = self._extract_ohmic(
            dictionary=config,
        )
        wiring_DC = self._extract_dcwiring(
            dictionary=config,
            ohmics=ohmics,
            connections=connections,
        )

        total_gates = [
            *connections.screening_gates,
            *connections.reservoir_gates,
            *connections.plunger_gates,
            *connections.barrier_gates,
        ]
        constraints = self._extract_voltage_constraints(
            map=config, total_gates=total_gates
        )
        return Config(
            ohmics=ohmics,
            wiring_DC=wiring_DC,
            groups=groups,
            screening_gates=connections.screening_gates,
            reservoir_gates=connections.reservoir_gates,
            plunger_gates=connections.plunger_gates,
            barrier_gates=connections.barrier_gates,
            constraints=constraints,
        )

    def _extract_adjacency(
        self,
        map: dict[str, dict[str, dict[str, str]]],
        total_gates: list,
    ) -> Adjacency:
        """Given the available gates loaded from the config, lets generate the adjacency matrix for this device."""
        num_gates = len(total_gates)
        total_gate_names = [gate.name for gate in total_gates]
        adjacency = np.array(np.zeros((num_gates, num_gates)), dtype=np.bool)
        name = "adjacency"
        if name not in map:
            msg = f"Expected to find {name} in {map} but the config did not contain the entry"
            raise IndexError(msg)
        adjacency_map = map[name]
        for i, gate in enumerate(total_gate_names):
            if gate not in adjacency_map:
                continue
            adjacent_matches = adjacency_map[gate]
            if not isinstance(adjacent_matches, str):
                msg = f"Expected the type of the adjacent matches in the config to be strings, but got{type(adjacent_matches)} instead."
                raise TypeError(msg)
            gate_names = adjacent_matches.split(";")
            for j, gate in enumerate(total_gate_names):
                if gate not in gate_names:
                    continue
                adjacency[i, j] = True
                adjacency[j, i] = True

        return Adjacency(
            matrix=adjacency,
            indexes=total_gates,
        )

    def _extract_voltage_constraints(
        self,
        map: dict[str, "Any"],
        total_gates: list,
    ) -> VoltageConstraints:
        """Given the available gates and the config lets produce a voltage constraints object."""
        adjacency = self._extract_adjacency(map=map, total_gates=total_gates)
        name = "max_safe_diff"
        if name not in map:
            msg = f"Expected to find index {name} in the config but only found available indexes {map.keys()}"
            raise IndexError(msg)
        max_safe_diff = map[name]
        if not isinstance(max_safe_diff, float):
            msg = f"Expected the type of the max safe diff to be a float, but got {type(max_safe_diff)} instead."
            raise TypeError(msg)
        name = "safe_voltage_bounds"
        if name not in map:
            msg = f"Expected to find index {name} in the config but only found available indexes {map.keys()}"
            raise IndexError(msg)
        bounds = map[name]
        if not isinstance(bounds, list):
            msg = f"Expected the type of the bounds to be a list, but got {type(bounds)} instead."
            raise TypeError(msg)
        for bound in bounds:
            if not isinstance(bound, float):
                msg = f"Expected the type of a bound inside the bounds to be float, but got {type(bounds)} instead."
                raise TypeError(msg)
        if len(bounds) != 2:
            msg = f"Expected two bounds stored for the max and min, but got {len(bounds)} bounds instead."
            raise ValueError(msg)
        return VoltageConstraints(
            adjacency=adjacency,
            max_safe_diff=max_safe_diff,
            bounds=tuple(bounds),
        )

    def _extract_dcwiring(
        self,
        dictionary: dict[str, dict[str, dict[str, float] | None]],
        ohmics: Ohmics,
        connections: StandardConfigConnections,
    ) -> Impedances:
        """Extracts a dcwiring from a dictionary.

        Args:
            dictionary : the dictionary that is being extracted from
            ohmics : the list of ohmics
        Returns:
            a list of Impedance objects instatiated
        Raises:
            TypeError if the dcwiring loaded from the config are not of type string
        """
        outs: list[Impedance] = []
        impedances = dictionary["wiringDC"]
        for key in list(impedances.keys()):
            values = impedances[key]
            if values is None:
                continue
            if key in [ohmic.name for ohmic in ohmics]:
                outs.append(
                    Impedance(
                        connection=Ohmic(key),
                        resistance=values["resistance"],
                        capacitance=values["capacitance"],
                    )
                )
            else:
                if key in [pgate.name for pgate in connections.plunger_gates]:
                    gt = PlungerGate
                elif key in [pgate.name for pgate in connections.barrier_gates]:
                    gt = BarrierGate
                elif key in [pgate.name for pgate in connections.screening_gates]:
                    gt = ScreeningGate
                elif key in [pgate.name for pgate in connections.reservoir_gates]:
                    gt = ReservoirGate
                else:
                    msg = "Cannot use that gate"
                    raise TypeError(msg)
                outs.append(
                    Impedance(
                        connection=gt(key),
                        resistance=values["resistance"],
                        capacitance=values["capacitance"],
                    )
                )
        return Impedances(outs)

    @overload
    def _extract_gates(
        self,
        dictionary: dict[str, str],
        gate_type: type["BarrierGate"],
    ) -> "BarrierGates": ...

    @overload
    def _extract_gates(
        self,
        dictionary: dict[str, str],
        gate_type: type["PlungerGate"],
    ) -> "PlungerGates": ...

    @overload
    def _extract_gates(
        self,
        dictionary: dict[str, str],
        gate_type: type["ReservoirGate"],
    ) -> "ReservoirGates": ...

    @overload
    def _extract_gates(
        self,
        dictionary: dict[str, str],
        gate_type: type["ScreeningGate"],
    ) -> "ScreeningGates": ...

    def _extract_gates(
        self,
        dictionary: dict[str, str],
        gate_type,
    ) -> "UsefulGates":
        """Extracts a gate from a dictionary where the gatetype is a key.

        Performs the tranformation from string index by ";" to list

        Args:
            dictionary : the dictionary that is being extracted from
            gate_type : the gatetype name / key to the dictionary

        Returns:
            a list of Gate objects instatiated

        Raises:
            TypeError if the gates loaded from the config are not of type string
        """
        for key, value in CONFIG_TO_GATE_PRIMITIVES.items():
            if value == gate_type:
                gatetype = key
                break
        else:
            msg = f"Invalid gate type {gate_type}"
            raise TypeError(msg)

        gates = dictionary[gatetype]
        if not isinstance(gates, str):
            msg = f"{gatetype} needs to be a str in config, delimitted by ;"
            raise TypeError(msg)
        return Gates([gate_type(gate) for gate in gates.split(";")])

    def _extract_standard_config_connections(
        self,
        config: dict[str, str],
    ) -> StandardConfigConnections:
        """Extracts the standard connections from a config dictionary.

        Args:
            config : the dictionary that is being extracted from

        Returns:
            a dictionary of the standard connections
        """
        screening_gates = self._extract_gates(
            dictionary=config,
            gate_type=ScreeningGate,
        )
        reservoir_gates = self._extract_gates(
            dictionary=config,
            gate_type=ReservoirGate,
        )
        plunger_gates = self._extract_gates(
            dictionary=config,
            gate_type=PlungerGate,
        )
        barrier_gates = self._extract_gates(
            dictionary=config,
            gate_type=BarrierGate,
        )
        return StandardConfigConnections(
            screening_gates=screening_gates,
            reservoir_gates=reservoir_gates,
            plunger_gates=plunger_gates,
            barrier_gates=barrier_gates,
            ohmics=Ohmics([]),
        )

    def _extract_ohmic(
        self,
        dictionary: dict[str, str],
    ) -> Ohmics:
        """Extracts a ohmic from a dictionary where the gatetype is a key.

        Performs the tranformation from string index by ";" to list

        Args:
            dictionary : the dictionary that is being extracted from
            gatetype : the gatetype name / key to the dictionary

        Returns:
            a list of Gate objects instatiated

        Raises:
            TypeError if the ohmic loaded from the config are not of type string
        """
        gatetype = "Ohmics"
        gates = dictionary[gatetype]
        if not isinstance(gates, str):
            msg = f"{gatetype} needs to be a str in config, delimitted by ;"
            raise TypeError(msg)
        return Ohmics([Ohmic(gate) for gate in gates.split(";")])

    def _extract_order(
        self,
        dictionary: dict[str, str],
        connections: StandardConfigConnections,
    ) -> "BaseConnections":
        """Extracts a gate from a dictionary.

        Performs the tranformation from string index by ";" to list

        Args:
            dictionary : the dictionary that is being extracted from
            connections : the various connections and their classes

        Returns:
            a list of Gate objects instatiated

        Raises:
            TypeError if the Order loaded from the config are not of type string
        """
        config = dictionary["Order"]
        if not isinstance(config, str):
            msg = "Order needs to be a str in config, delimitted by ;"
            raise TypeError(msg)
        raworder = config.split(";")
        order: list[Connection] = []
        for gate in raworder[1:-1]:
            if gate in [sgate.name for sgate in connections.screening_gates]:
                gatetype = ScreeningGate
            elif gate in [bgate.name for bgate in connections.barrier_gates]:
                gatetype = BarrierGate
            elif gate in [pgate.name for pgate in connections.plunger_gates]:
                gatetype = PlungerGate
            else:  # will be ReservoirGate
                gatetype = ReservoirGate
            order.append(gatetype(gate))
        order.insert(0, Ohmic(raworder[0]))
        order.append(Ohmic(raworder[-1]))
        return BaseConnections(order)

File: cpp/src/physics/config/ConfigManipulations.cpp
#include "falcon_core/physics/config/ConfigManipulations.hpp"

#include <stdexcept>

#include "falcon_core/physics/config/core/Group.hpp"
#include "falcon_core/physics/config/core/StandardConfigConnections.hpp"
#include "falcon_core/physics/device_structures/Connections.hpp"
#include "falcon_core/physics/device_structures/Impedances.hpp"
namespace falcon_core {
namespace physics {
namespace config {
using core::StandardConfigConnections;

std::vector<std::string> split_on_semicolon(const std::string& s) {
  std::vector<std::string> result;
  std::string::size_type   start = 0, end;
  while ((end = s.find(';', start)) != std::string::npos) {
    result.push_back(s.substr(start, end - start));
    start = end + 1;
  }
  result.push_back(s.substr(start));
  return result;
}
ConfigManipulations::ConfigManipulations() = default;
core::ConfigSP ConfigManipulations::unpack_device_config(
    const YAML::Node& config) const {
  generic::Map<autotuner_interfaces::names::Gname, core::Group> groups;
  core::StandardConfigConnectionsSP                             connections =
      _extract_standard_config_connections(config);
  for (const auto& group_pair : config["groups"]) {
    auto key               = group_pair.first.as<std::string>();
    auto value             = group_pair.second;
    auto group_connections = _extract_standard_group_config_connections(value);
    auto order =
        _extract_order(value["Order"].as<std::string>(), group_connections);
    core::GroupSP new_group = std::make_shared<core::Group>(
        std::make_shared<autotuner_interfaces::names::Channel>(key),
        value["NumDots"].as<int>(),
        group_connections->screening_gates(),
        group_connections->reservoir_gates(),
        group_connections->plunger_gates(),
        group_connections->barrier_gates(),
        order);
    groups.insert(std::make_shared<autotuner_interfaces::names::Gname>(key),
                  new_group);
  }
  auto ohmics    = _extract_ohmics(config["Ohmics"].as<std::string>());
  auto wiring_DC = _extract_dcwiring(config, ohmics, connections);

  device_structures::Connections total_gates;
  total_gates.insert(total_gates.end(),
                     connections->screening_gates()->begin(),
                     connections->screening_gates()->end());
  total_gates.insert(total_gates.end(),
                     connections->reservoir_gates()->begin(),
                     connections->reservoir_gates()->end());
  total_gates.insert(total_gates.end(),
                     connections->plunger_gates()->begin(),
                     connections->plunger_gates()->end());
  total_gates.insert(total_gates.end(),
                     connections->barrier_gates()->begin(),
                     connections->barrier_gates()->end());

  auto constraints = _extract_voltage_constraints(
      config, std::make_shared<device_structures::Connections>(total_gates));

  return std::make_shared<core::Config>(
      connections->screening_gates(),
      connections->plunger_gates(),
      ohmics,
      connections->barrier_gates(),
      connections->reservoir_gates(),
      std::make_shared<
          generic::Map<autotuner_interfaces::names::Gname, core::Group>>(
          groups),
      wiring_DC,
      constraints);
}

core::AdjacencySP ConfigManipulations::_extract_adjacency(
    const YAML::Node&                       map,
    const device_structures::ConnectionsSP& total_gates) const {
  if (!total_gates) {
    throw std::invalid_argument(
        "ConfigManipulations: The total_gates cannot be null.");
  }
  size_t               num_gates = total_gates->size();
  generic::FArray<int> adjacency =
      *generic::FArray<int>::zeros({num_gates, num_gates});
  std::vector<std::string> total_gate_names;
  for (const auto& gate : *total_gates)
    total_gate_names.push_back(gate->name());

  if (!map["adjacency"])
    throw std::runtime_error("Expected to find adjacency in config");
  auto adjacency_map = map["adjacency"];
  for (size_t i = 0; i < total_gate_names.size(); ++i) {
    auto gate = total_gate_names[i];
    if (!adjacency_map[gate]) continue;
    auto adjacent_matches = adjacency_map[gate].as<std::string>();
    auto gate_names       = split_on_semicolon(adjacent_matches);
    for (size_t j = 0; j < total_gate_names.size(); ++j) {
      if (std::find(gate_names.begin(),
                    gate_names.end(),
                    total_gate_names[j]) != gate_names.end()) {
        adjacency(i, j) = 1;
        adjacency(j, i) = 1;
      }
    }
  }
  return std::make_shared<core::Adjacency>(adjacency.data(), total_gates);
}

core::VoltageConstraintsSP ConfigManipulations::_extract_voltage_constraints(
    const YAML::Node&                       map,
    const device_structures::ConnectionsSP& total_gates) const {
  auto adjacency = _extract_adjacency(map, total_gates);
  if (!map["max_safe_diff"])
    throw std::runtime_error("Expected max_safe_diff in config");
  double max_safe_diff = map["max_safe_diff"].as<double>();
  if (!map["safe_voltage_bounds"])
    throw std::runtime_error("Expected safe_voltage_bounds in config");
  auto                bounds_node = map["safe_voltage_bounds"];
  std::vector<double> bounds;
  for (const auto& b : bounds_node) bounds.push_back(b.as<double>());
  if (bounds.size() != 2)
    throw std::runtime_error("Expected two bounds for voltage");
  return std::make_shared<core::VoltageConstraints>(
      adjacency, max_safe_diff, std::make_pair(bounds[0], bounds[1]));
}

device_structures::ImpedancesSP ConfigManipulations::_extract_dcwiring(
    const YAML::Node&                        map,
    const device_structures::ConnectionsSP&  ohmics,
    const core::StandardConfigConnectionsSP& connections) const {
  if (!ohmics || !connections) {
    throw std ::runtime_error(
        "ConfigManipulations: The ohmics and the connecations "
        "cannot be null.");
  }
  device_structures::Impedances outs;
  auto                          impedances = map["wiringDC"];
  for (const auto& entry : impedances) {
    std::string                     key    = entry.first.as<std::string>();
    auto                            values = entry.second;
    device_structures::ConnectionSP gt;
    device_structures::ConnectionSP ohmic =
        device_structures::Connection::Ohmic(key);
    if (connections->has_ohmic(ohmic)) {
      gt = ohmic;
    } else {
      device_structures::ConnectionSP plunger_gate =
          device_structures::Connection::PlungerGate(key);
      if (connections->has_plunger_gate(plunger_gate)) {
        gt = plunger_gate;
      } else {
        device_structures::ConnectionSP barrier_gate =
            device_structures::Connection::BarrierGate(key);
        if (connections->has_barrier_gate(barrier_gate)) {
          gt = barrier_gate;
        } else {
          device_structures::ConnectionSP screening_gate =
              device_structures::Connection::ScreeningGate(key);
          if (connections->has_screening_gate(screening_gate)) {
            gt = screening_gate;
          } else {
            device_structures::ConnectionSP reservoir_gate =
                device_structures::Connection::ReservoirGate(key);
            if (connections->has_reservoir_gate(reservoir_gate)) {
              gt = reservoir_gate;
            } else {
              throw std::runtime_error(
                  "ConfigManipulations: Cannot use that connection : " + key);
            }
          }
        }
      }
    }

    outs.push_back(std::make_shared<device_structures::Impedance>(
        gt,
        values["resistance"].as<double>(),
        values["capacitance"].as<double>()));
  }
  return std::make_shared<device_structures::Impedances>(outs);
}

device_structures::ConnectionsSP ConfigManipulations::_extract_barrier_gates(
    const std::string raw) const {
  auto gate_names = split_on_semicolon(raw);
  std::vector<device_structures::ConnectionSP> barriergates;
  for (const auto& name : gate_names)
    barriergates.push_back(device_structures::Connection::BarrierGate(name));
  return std::make_shared<device_structures::Connections>(barriergates);
}
device_structures::ConnectionsSP ConfigManipulations::_extract_plunger_gates(
    const std::string raw) const {
  auto gate_names = split_on_semicolon(raw);
  std::vector<device_structures::ConnectionSP> plungergates;
  for (const auto& name : gate_names)
    plungergates.push_back(device_structures::Connection::PlungerGate(name));
  return std::make_shared<device_structures::Connections>(plungergates);
}
device_structures::ConnectionsSP ConfigManipulations::_extract_reservoir_gates(
    const std::string raw) const {
  auto gate_names = split_on_semicolon(raw);
  std::vector<device_structures::ConnectionSP> reservoirgates;
  for (const auto& name : gate_names)
    reservoirgates.push_back(
        device_structures::Connection::ReservoirGate(name));
  return std::make_shared<device_structures::Connections>(reservoirgates);
}
device_structures::ConnectionsSP ConfigManipulations::_extract_screening_gates(
    const std::string raw) const {
  auto gate_names = split_on_semicolon(raw);
  std::vector<device_structures::ConnectionSP> screeninggates;
  for (const auto& name : gate_names)
    screeninggates.push_back(
        device_structures::Connection::ScreeningGate(name));
  return std::make_shared<device_structures::Connections>(screeninggates);
}
device_structures::ConnectionsSP ConfigManipulations::_extract_ohmics(
    const std::string raw) const {
  auto gate_names = split_on_semicolon(raw);
  std::vector<device_structures::ConnectionSP> ohmics;
  for (const auto& name : gate_names)
    ohmics.push_back(device_structures::Connection::Ohmic(name));
  return std::make_shared<device_structures::Connections>(ohmics);
}

core::StandardConfigConnectionsSP
ConfigManipulations::_extract_standard_group_config_connections(
    const YAML::Node& config) const {
  auto screening_gates =
      _extract_screening_gates(config["ScreeningGates"].as<std::string>());
  auto reservoir_gates =
      _extract_reservoir_gates(config["ReservoirGates"].as<std::string>());
  auto plunger_gates =
      _extract_plunger_gates(config["PlungerGates"].as<std::string>());
  auto barrier_gates =
      _extract_barrier_gates(config["BarrierGates"].as<std::string>());
  return std::make_shared<core::StandardConfigConnections>(
      screening_gates,
      reservoir_gates,
      plunger_gates,
      barrier_gates,
      std::make_shared<device_structures::Connections>());
}

core::StandardConfigConnectionsSP
ConfigManipulations::_extract_standard_config_connections(
    const YAML::Node& config) const {
  auto screening_gates =
      _extract_screening_gates(config["ScreeningGates"].as<std::string>());
  auto reservoir_gates =
      _extract_reservoir_gates(config["ReservoirGates"].as<std::string>());
  auto plunger_gates =
      _extract_plunger_gates(config["PlungerGates"].as<std::string>());
  auto barrier_gates =
      _extract_barrier_gates(config["BarrierGates"].as<std::string>());
  auto ohmics = _extract_ohmics(config["Ohmics"].as<std::string>());
  return std::make_shared<core::StandardConfigConnections>(
      screening_gates, reservoir_gates, plunger_gates, barrier_gates, ohmics);
}

device_structures::ConnectionsSP ConfigManipulations::_extract_order(
    const std::string                        raw,
    const core::StandardConfigConnectionsSP& connections) const {
  if (!connections) {
    throw std::invalid_argument(
        "ConfigManipulations: The connections cannot be null.");
  }
  auto                             raworder = split_on_semicolon(raw);
  device_structures::ConnectionsSP order =
      std::make_shared<device_structures::Connections>();
  for (size_t i = 1; i + 1 < raworder.size(); ++i) {
    const auto&                     gate = raworder[i];
    device_structures::ConnectionSP realGate =
        device_structures::Connection::ScreeningGate(gate);
    if (connections->has_screening_gate(realGate)) {
      order->push_back(realGate);
    } else {
      device_structures::ConnectionSP realGate =
          device_structures::Connection::BarrierGate(gate);
      if (connections->has_barrier_gate(realGate)) {
        order->push_back(realGate);
      } else {
        device_structures::ConnectionSP realGate =
            device_structures::Connection::PlungerGate(gate);
        if (connections->has_plunger_gate(realGate)) {
          order->push_back(realGate);
        } else {
          device_structures::ConnectionSP realGate =
              device_structures::Connection::ReservoirGate(gate);
          if (connections->has_reservoir_gate(realGate)) {
            order->push_back(realGate);
          } else
            throw std::runtime_error("Unexpected gate name found of " + gate +
                                     " does not belong in config.");
        }
      }
    }
  }
  auto ohmic_conn = device_structures::Connection::Ohmic(raworder.front());
  std::vector<device_structures::ConnectionSP> temp{ohmic_conn};
  order->insert(order->begin(), temp.begin(), temp.end());
  order->push_back(device_structures::Connection::Ohmic(raworder.back()));
  return order;
}

}  // namespace config
}  // namespace physics
}  // namespace falcon_core
